prometheus.lua

```diff
@@ -320,260 +140,374 @@ local function check_metric_and_label_names(metric_name, label_names)
-  self.type[name] = "histogram"
 
-  self.buckets[name] = buckets or DEFAULT_BUCKETS
-  self.bucket_format[name] = construct_bucket_format(self.buckets[name])
+  -- sum
+  _, err = c:incr(keys[2], value, 0)
+  if err then
+    self._log_error(err)
+  end
 
-  return Histogram:new{name=name, label_names=label_names, prometheus=self}
+  local seen = false
+  -- check in reverse order, otherwise we will always
+  -- need to traverse the whole table.
+  for i=self.bucket_count, 1, -1 do
+    if value <= self.bucket[i] then
+      _, err = c:incr(keys[2+i], 1, 0)
+      if err then
+        self._log_error(err)
+      end
+      seen = true
+    elseif seen then
+      break
+    end
+  end
+  -- inf
+  _, err = c:incr(keys[self.bucket_count+3], 1, 0)
+  if err then
+    self._log_error(err)
+  end
 end
 
--- Set a given dictionary key.
--- This overwrites existing values, so it should only be used when initializing
--- metrics or when explicitely overwriting the previous value of a metric.
-function Prometheus:set_key(key, value)
-  local ok, err = self.dict:safe_set(key, value)
-  if not ok then
-    self:log_error_kv(key, value, err)
+local function reset(self)
+  if self.parent.resty_counter_used then
+    ngx.log(ngx.INFO, "waiting ", self.parent.sync_interval, "s for counter to sync")
+    ngx.sleep(self.parent.sync_interval)
   end
+
+  local keys = self._dict:get_keys(0)
+  local name_prefix = self.name .. "{"
+  local name_prefix_length = #name_prefix
+
+  for _, key in ipairs(keys) do
+    local value, err = self._dict:get(key)
+    if value then
+      -- with out labels equal, or with labels and the part before { equals
+      if key == self.name or name_prefix == string.sub(key, 1, name_prefix_length) then
+        _, err = self._dict:safe_set(key, nil)
+        if err then
+          self._log_error("Error resetting '", key, "': ", err)
+        end
+      end
+    else
+      self._log_error("Error getting '", key, "': ", err)
+    end
+  end
+
+  -- clean up lookup table as well
+  self.lookup = {}
 end
 
--- Increment a given metric by `value`.
---
--- Args:
---   name: (string) short metric name without any labels.
---   label_names: (array) a list of label keys.
---   label_values: (array) a list of label values.
---   value: (number) value to add (a negative value when you need to decrease
---     the value of the gauge). Optional, defaults to 1.
-function Prometheus:inc(name, label_names, label_values, value)
-  local key = full_metric_name(name, label_names, label_values)
-  if value == nil then value = 1 end
-
-  local newval, err = self.dict:incr(key, value)
-  if newval then
-    return
+-- ends metrics functions
+
+function Prometheus.init(dict_name, prefix, sync_interval)
+  local self = setmetatable({}, mt)
+  dict_name = dict_name or "prometheus_metrics"
+  self.dict_name = dict_name
+  self.dict = ngx.shared[dict_name]
+  if self.dict == nil then
+    error("Dictionary '" .. dict_name .. "' does not seem to exist. " ..
```

Comment:
this function is expected to be called in init_by_lua phase, since self.dict being nil will cause prometheus library to not function at all, so we can fail hard here. then later there's also no need to check if self.dict is nil or not in hot path. let me know you thoughts on this change : )